File: algos/insertion_sort_list.py
class ListNode:
	def __init__(self, x):
		self.val = x
		self.next = None

# given the head of a linked list,
# perform an insertion sort on the linked list
def insertion_sort(head):
	if not head:
		return None
		
	node1 = head.next
	while node1:
		node2 = head
		while node2.val < node1.val:
			node2 = node2.next
			
		oldval = node1.val
		while node2 != node1:
			oldval, node2.val = node2.val, oldval
			node2 = node2.next
		node1.val = oldval
			
		node1 = node1.next
			
	return head

File: sort/insertionSortList.java
public class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }
}

public class insertionSortList {
    public ListNode sort(ListNode head) {
        if (head == null)
            return null;
            
        ListNode node1;
        ListNode node2;
        int tmp;
        int oldval;
            
        node1 = head;
        while (node1 != null) {
            node2 = head;
            while (node2.val < node1.val)
                node2 = node2.next;
            
            oldval = node1.val;
            while (node2 != node1) {
                tmp = oldval;
                oldval = node2.val;
                node2.val = tmp;
                
                node2 = node2.next;
            }
            node1.val = oldval;
            
            node1 = node1.next;
        }
        
        return head;
    }
}
